### DISCUSSION TITLE
About post in an Action

### DISCUSSION THREAD

**Discussion by DaHuoKolmostar** [community/community#26504] [Actions]
<p>I noticed in the <a href="https://github.com/actions/checkout/blob/master/action.yml" rel="nofollow noopener">action.yaml</a> of actions/checkout@v2 , under the <strong>runs</strong> command, there is a section named “_ <strong>post</strong> _”, which is not documented in <a href="https://help.github.com/en/actions" rel="nofollow noopener">https://help.github.com/en/actions</a>. I guess the “post action” is called after all other steps, before the job finishes.</p>
<pre><code>name: 'Checkout'
description: 'Checkout a Git repository at a particular version'
inputs:
  repository:
    description: 'Repository name with owner. For example, actions/checkout'
    default: ${{ github.repository }}
  ref:
    description: &gt;
      The branch, tag or SHA to checkout. When checking out the repository that
      triggered a workflow, this defaults to the reference or SHA for that
      event. Otherwise, defaults to `master`.
  token:
    description: &gt;
      Auth token used to fetch the repository. The token is stored in the local
      git config, which enables your scripts to run authenticated git commands.
      The post-job step removes the token from the git config.
    default: ${{ github.token }}
  persist-credentials:
    description: 'Whether to persist the token in the git config'
    default: true
  path:
    description: 'Relative path under $GITHUB_WORKSPACE to place the repository'
  clean:
    description: 'Whether to execute `git clean -ffdx &amp;&amp; git reset --hard HEAD` before fetching'
    default: true
  fetch-depth:
    description: 'Number of commits to fetch. 0 indicates all history.'
    default: 1
  lfs:
    description: 'Whether to download Git-LFS files'
    default: false
runs:
  using: node12
  main: dist/index.js
  post: dist/index.js
</code></pre>
<p>Can anyone help me explain how this works? Is it possible to use “post” in self-defined actions, and if so how to use it? Will it also be utilized in Docker actions?</p>

**Comment by BrightRan** [comment, marked as answer]
<p>This is related to the new feature <strong>Wrapper action</strong> which is being introduced. When you want to run some cleanup operations at the end of the job execution to cleanup any states changes caused by your action, you can use <strong>post</strong> in the <strong>action.yaml</strong> of your action.</p>
<p>The runner will add a step to its job’s steps list when the current running action has <strong>post</strong> defined in the <strong>action.yaml</strong> , the post job step will be added with condition set to <strong>always()</strong> by default and can be overridden via <strong>post-if</strong>.</p>
<p>The post job cleanup action will always get a chance to run as long as the entire job haven’t exceeded job timeout or job cancel timeout. And the action’s post job execution will share a state <strong>Dictionary&lt;string, string&gt;</strong> with only the action’s main execution, so the main execution can save some context for the post execution, like what needs to be cleanup.</p>
<p>As I mentioned above, this new feature is being introduced, currently there is not released documentation for it.</p>

**Comment by DaHuoKolmostar** [comment]
<p>Thanks BrightRan! Really looking forward to this new feature!</p>

**Comment by sjurgis** [comment]
<p>Is there any way to cleanup after workflow finishes?</p>
<p>Our’s authorizes into remote system and creates an environment there. After tests finish (fail or succeed) we need to destroy the environment (kinda crucial).</p>

**Comment by davidkarlsen** [comment]
<p>is this available for docker based actions as well? How do you access the context in these cases?</p>
